lib/src/api/opt/resource.rs

```diff
@@ -1,58 +1,115 @@
-use crate::api::{err::Error, Result};
-use crate::sql::{self, Array, Edges, Id, Object, Table, Thing, Value};
-use crate::syn;
+use crate::{
+	api::{err::Error, Result},
+	Object, RecordId, RecordIdKey, Value,
+};
 use std::ops::{self, Bound};
+use surrealdb_core::sql::{
+	Edges as CoreEdges, IdRange as CoreIdRange, Table as CoreTable, Thing as CoreThing,
+};
+
+#[cfg(any(feature = "protocol-ws", feature = "protocol-http"))]
+use surrealdb_core::sql::Value as CoreValue;
+
+/// A wrapper type to assert that you ment to use a string as a table name.
+///
+/// To prevent some possible errors, by defauit [`IntoResource`] does not allow `:` in table names
+/// as this might be an indication that the user might have intended to use a record id instead.
+/// If you wrap your table name string in this tupe the [`IntoResource`] trait will accept any
+/// table names.
+#[derive(Debug)]
+pub struct Table<T>(pub T);
+
+impl<T> Table<T>
+where
+	T: Into<String>,
+{
+	pub(crate) fn into_core(self) -> CoreTable {
+		let mut t = CoreTable::default();
+		t.0 = self.0.into();
+		t
+	}
+
+	/// Add a range of keys to the table.
+	pub fn with_range<R>(self, range: R) -> QueryRange
+	where
+		KeyRange: From<R>,
+	{
+		let range = KeyRange::from(range);
+		let res = CoreIdRange {
+			beg: range.start.map(RecordIdKey::into_inner),
+			end: range.end.map(RecordIdKey::into_inner),
+		};
+		let res = CoreThing::from((self.0.into(), res));
+		QueryRange(res)
+	}
+}
+
+transparent_wrapper!(
+	/// A table range.
+	#[derive( Clone, PartialEq)]
+	pub struct QueryRange(CoreThing)
+);
+
+transparent_wrapper!(
+	/// A query edge
+	#[derive( Clone, PartialEq)]
+	pub struct Edge(CoreEdges)
+);
 
 /// A database resource
-#[derive(Debug)]
+///
+/// A resource is a location, or a range of locations, from which data can be fetched.
+#[derive(Debug, Clone, PartialEq)]
+#[non_exhaustive]
 pub enum Resource {
 	/// Table name
-	Table(Table),
+	Table(String),
 	/// Record ID
-	RecordId(Thing),
+	RecordId(RecordId),
 	/// An object
 	Object(Object),
 	/// An array
-	Array(Array),
+	Array(Vec<Value>),
 	/// Edges
-	Edges(Edges),
+	Edge(Edge),
+	/// A range of id's on a table.
+	Range(QueryRange),
 }
 
 impl Resource {
-	pub(crate) fn with_range(self, range: Range<Id>) -> Result<sql::Thing> {
+	/// Add a range to the resource, this only works if the resource is a table.
+	pub fn with_range(self, range: KeyRange) -> Result<Self> {
```

Comment:
Does this method really need to be public?

Reply:
It doesn't have to be but it is convenient as it allows you add a range to a manually created resource.